apiv6/kentikapi/httputil/httputil.go

```diff
@@ -39,8 +37,141 @@ func NewRetryingClient(cfg ClientConfig) *http.Client {
 	if cfg.RetryWaitMax != nil {
 		c.RetryWaitMax = *cfg.RetryWaitMax
 	}
-	c.CheckRetry = retryablehttp.ErrorPropagatedRetryPolicy
+	c.CheckRetry = makeRetryPolicy(cfg.RetryableStatusCodes, cfg.RetryableMethods)
 	c.ErrorHandler = retryablehttp.PassthroughErrorHandler
 
 	return c.StandardClient()
 }
+
+// ClientConfig holds configuration for retrying client.
+type ClientConfig struct {
+	HTTPClient           *http.Client
+	RetryMax             *int
+	RetryWaitMin         *time.Duration
+	RetryWaitMax         *time.Duration
+	RetryableStatusCodes []int
+	RetryableMethods     []string
+}
+
+func (cfg *ClientConfig) FillDefaults() {
+	if cfg.HTTPClient == nil {
+		cfg.HTTPClient = defaultHTTPClient()
+	}
+
+	if len(cfg.RetryableStatusCodes) == 0 {
+		cfg.RetryableStatusCodes = []int{429, 500, 502, 503, 504}
+	}
+
+	if len(cfg.RetryableMethods) == 0 {
+		cfg.RetryableMethods = []string{
+			http.MethodGet, http.MethodHead, http.MethodPost, http.MethodPut, http.MethodPatch, http.MethodDelete,
+			http.MethodConnect, http.MethodOptions, http.MethodTrace,
+		}
+	}
+}
+
+func defaultHTTPClient() *http.Client {
+	return &http.Client{
+		Transport: &http.Transport{
+			Proxy: http.ProxyFromEnvironment,
+			DialContext: (&net.Dialer{
+				Timeout:   30 * time.Second,
+				KeepAlive: 30 * time.Second,
+			}).DialContext,
+			ForceAttemptHTTP2:     true,
+			MaxIdleConns:          100,
+			IdleConnTimeout:       90 * time.Second,
+			TLSHandshakeTimeout:   10 * time.Second,
+			ExpectContinueTimeout: 1 * time.Second,
+		},
+	}
+}
+
+// makeRetryPolicy creates customized retry policy.
+// Its implementation is based on retryablehttp.ErrorPropagatedRetryPolicy.
+// Retry policy function returns true if the request should be retried.
+func makeRetryPolicy(statusCodes []int, methods []string) retryablehttp.CheckRetry {
+	statusCodesSet := makeIntSet(statusCodes)
+	methodsSet := makeStringSet(methods)
+	redirectsErrorRe := redirectsErrorRegexp()
+	schemeErrorRe := schemeErrorRegexp()
+
+	return func(ctx context.Context, resp *http.Response, err error) (bool, error) {
+		// do not retry on context.Canceled or context.DeadlineExceeded
+		if ctx.Err() != nil {
+			return false, ctx.Err()
+		}
+
+		if !isRequestRetrayable(resp.Request, methodsSet) {
+			return false, err
+		}
+
+		if err != nil {
+			// TODO(dfurman): unit test that branch
+			return isErrorRecoverable(err, redirectsErrorRe, schemeErrorRe)
+		}
+
+		if _, ok := statusCodesSet[resp.StatusCode]; ok {
+			return true, nil
+		}
+
+		return false, nil
+	}
+}
+
+func isRequestRetrayable(r *http.Request, methodsSet map[string]struct{}) bool {
+	_, ok := methodsSet[r.Method]
+	return ok
+}
+
+func isErrorRecoverable(err error, redirectsErrorRe, schemeErrorRe *regexp.Regexp) (bool, error) {
```

Comment:
Telling errors one from another by parsing error message can prove brittle as the message can change without notice, is there maybe some more robust way?

Reply:
Fixed in https://github.com/kentik/community_sdk_golang/pull/42